### DISCUSSION TITLE
Can I add customized arguments to an agent's "step" method?

### DISCUSSION THREAD

**Discussion by cylin928** [mesa/mesa#1713] [Q&A] [answered]
Hi, 
Can I add customized arguments to an agent's "step" method?
I tried it with the following code, in which I add an "a" argument to the step method.
What do you think the best way is to address a need like this?
Simply assign new data by updating agents' attributes before running "step()"?

Thank you very much.

`import mesa
class MoneyAgent(mesa.Agent):
    """An agent with fixed initial wealth."""

    def __init__(self, unique_id, model):
        # Pass the parameters to the parent class.
        super().__init__(unique_id, model)

        # Create the agent's attribute and set the initial values.
        self.wealth = 1
        self.d = {"sds": {}}

    def step(self, a):
        # The agent's step will go here.
        # For demonstration purposes we will print the agent's unique_id
        print(f"Hi, I am an agent, you can call me {str(self.unique_id)}. {a}")

def compute_gini(model):
    agent_wealths = [agent.wealth for agent in model.schedule.agents]
    x = sorted(agent_wealths)
    N = model.num_agents
    B = sum(xi * (N - i) for i, xi in enumerate(x)) / (N * sum(x))
    return 1 + (1 / N) - 2 * B

class MoneyModel(mesa.Model):
    """A model with some number of agents."""

    def __init__(self, N):
        self.num_agents = N
        # Create scheduler and assign it to the model
        self.schedule = mesa.time.RandomActivation(self)

        # Create agents
        for i in range(self.num_agents):
            a = MoneyAgent(i, self)
            # Add the agent to the scheduler
            self.schedule.add(a)

        self.datacollector = mesa.DataCollector(
            model_reporters={"Gini": compute_gini},
            agent_reporters={"Wealth": "wealth",
                             "d": "d"}
        )

    def step(self):
        """Advance the model by one step."""

        # The model's step will go here for now this will call the step method of each agent and print the agent's unique_id
        self.datacollector.collect(self)
        self.schedule.step(a="Philip")

model = MoneyModel(10)
for i in range(100):
    model.step()`

**Comment by rht** [comment, marked as answer]
From the snippet, it seems to me that `a`'s value doesn't vary across agents. As such, specifying it as a model attribute that is read by the agents should be sufficient. The agents can access the value via `self.model.a`.
